master/internal/rm/agentrm/fitting.go

```diff
@@ -150,25 +165,52 @@ func findDedicatedAgentFits(
 		return nil
 	}
 
-	var candidates candidateList
-	for _, agent := range agentsByNumSlots[candidateNumSlots] {
-		candidates = append(candidates, &fittingState{
-			Agent:        agent,
-			Score:        fittingMethod(req, agent),
-			HashDistance: hashDistance(req, agent),
-		})
-	}
+	// First try to find a fit among agents with homogenous sizes.
+	for _, group := range candidateGroups {
+		numSlotsAvailable := len(group.candidateList) * group.slotsPerCandidate
+		if numSlotsAvailable < req.SlotsNeeded {
+			continue
+		}
+		if req.SlotsNeeded%group.slotsPerCandidate != 0 {
+			continue
+		}
 
-	sort.Sort(candidates)
+		sort.Sort(group.candidateList)
+		numNodesNeeded := req.SlotsNeeded / group.slotsPerCandidate
+		return group.candidateList[:numNodesNeeded]
+	}
 
-	numContainers := req.SlotsNeeded / candidateNumSlots
-	slotsPerContainer := req.SlotsNeeded / numContainers
-	fits := candidates[:numContainers]
-	for _, c := range fits {
-		c.Slots = slotsPerContainer
+	// Then try to find a fit among agents with heterogeneous sizes.
+	if allowHeterogeneousAgentFits {
+		var fits []*fittingState
+
+		// Progressively fit against nodes from largest to smallest. This works when node sizes are
+		// powers of some k, since a consuming some size k**2 is equivalent to consuming k nodes of
+		// k size, but with more nodes left find a fit. But if sizes aren't powers of some k, it
+		// doesn't work; with sizes 4, 6, and 8, a 12 slot job won't fit since we don't backtrack
+		// after not finding a fit with 8 and 6.
```

Comment:
do we have some way to enforce this / warn if this assumption is violated?

Reply:
eh, maybe not a bad idea, but for now it is just in the docs (i doubt anyone would see a warning buried in master logs, anyway)